### ISSUE TITLE
Support Scala Native

### ISSUE THREAD

**Issue by olafurpg** [scalameta/scalafmt#1172]
It would be awesome to support the scalafmt cli on native to avoid JVM startup time. I managed to get it working with `publishLocal` a while ago and experienced ~30x speedup formatting a single file (1.5s on JVM vs 50ms on Native)

https://twitter.com/olafurpg/status/857559907876433920

The following milestones are now complete

- https://github.com/scalameta/scalafmt/pull/1170 upgrade to Scalameta v3.7 and
- https://github.com/scalameta/scalameta/pull/1529 cross-build Scalameta to Scala Native

We are only a few steps away from being able to properly support Scala Native. The missing pieces are

- [x] scalacheck native, blocked by https://github.com/rickynils/scalacheck/issues/396
- [x] paiges native, blocked by https://github.com/typelevel/paiges/pull/87
- [x] metaconfig native
- [ ] refactor scalafmt-cli to remove JVM stuff, or reimplement the cli from scratch (tempting, the code is not so great)
- [ ] change the config parser to a pure Scala implementation, I'm tempted to require .scalafmt.json for Native/JS since there is no official HOCON parser written in Scala

**Comment by virusdave** [comment]
Any updates on this?  Any remaining blockers people can help with?

**Comment by ssaavedra** [comment]
To be fair, some improvements are already in place: paiges is already available for scala-native and metaconfig seems to work for native too although it's not in Central.

I tried something different, that is AOT compilation with GraalVM's `native-image`, which would also avoid the VM warmup penalty and we would still enjoy Lightbend's config for HOCON.

Compilation worked fine (I did an `sbt cli/assembly` and then `native-image -jar scalafmt.jar`) but when running the ELF with something like `scalafmt $PWD/SomeClass.scala` execution halts due to a ` com.oracle.svm.core.jdk.UnsupportedFeatureError: Code that was considered unreachable by closed-world analysis was reached` with this [stacktrace][stacktrace].

I'm not sure if it's a faster approach to go towards scala-native or graalvm, but I'm raising the issue to consider it.

  [stacktrace]: https://gist.github.com/ssaavedra/1b3817ab3bdb55f8a9912fdddf181175

**Comment by ekrich** [comment]
Some update on this. I have ported the config library and published for JVM and have a few branches to add to a potential Scala Native 0.3.9 release. I have successfully run the bit of code used to parse `scalafmt` config files in Scala Native. 

https://github.com/ekrich/sconfig

PR to use this in metaconfig which is used by `scalafmt` - I think the default still use lightbend/config for JVM.
https://github.com/olafurpg/metaconfig/commit/b2363df06365eb418e9e9531434e5956a3a95c7b

**Comment by ekrich** [comment]
We now have the changes in Scala Native 0.3.x to allow [sconfig](https://github.com/ekrich/sconfig) to cross compile to Scala Native and run the code `metaconfig` uses to parse the `scalafmt` HOCON config file.

**Comment by larsrh** [comment]
Paiges available now.

**Comment by hepin1989** [comment]
Scala native 0.4.0 is just released https://scala-native.readthedocs.io/en/v0.4.0/changelog/0.4.0.html

**Comment by ekrich** [comment]
Status of `sconfig` - https://github.com/ekrich/sconfig/pull/139

**Comment by ekrich** [comment]
I have released a version of `java.time`, [`sjavatime`](https://github.com/ekrich/sconfig), and a new version of [`sconfig`](https://github.com/ekrich/sconfig) that should be ready and not be blocking this anymore.

**Comment by tgodzik** [comment]
I think biggest blocking issue is scalameta currently :/ The scala native support for removed some time ago and we need to add it back.

**Comment by tgodzik** [comment]
Ok, I almost got something working : https://github.com/scalameta/scalameta/pull/2261

**Comment by tgodzik** [comment]
There is a compilation error though, but we are looking into this.

**Comment by ekrich** [comment]
I wanted to point out this although I am not sure if it has been solved or not yet - https://github.com/scala-native/scala-native/issues/1575 

The compilation error should be fixed in master - https://github.com/scala-native/scala-native/commit/0ded408cf0acebfac12049f875675a6efc31544f

I see you have provided a workaround in `scalameta` here: https://github.com/scalameta/scalameta/commit/2db5f934bf45d308a0830318658963f895b39e94#diff-865573f4d2cf3a17bdfd191d2665095bf22d743bc9433628d09b677103786d54L270-L281

**Comment by ekrich** [comment]
@tgodzik Let me know if you need any help.

**Comment by tgodzik** [comment]
I am down to to errors:
```
[error] Found 2 missing definitions while linking
[error] Not found Top(scopt.Read$)
[error]         at /home/tgodzik/Documents/scalafmt/scalafmt-cli/shared/src/main/scala/org/scalafmt/cli/CliArgParser.scala:61
[error]         at /home/tgodzik/Documents/scalafmt/scalafmt-cli/shared/src/main/scala/org/scalafmt/cli/CliArgParser.scala:64
....
[error] Not found Top(org.scalafmt.interfaces.Scalafmt$)
[error]         at /home/tgodzik/Documents/scalafmt/scalafmt-cli/shared/src/main/scala/org/scalafmt/cli/ScalafmtDynamicRunner.scala:26
```

I can't figure out why scopt is problematic since that is released for native. I do have to work on couple of things still, but I should get something compilable soon.

**Comment by tgodzik** [comment]
Weird thing, if I do `import scopt.Read` by hand it works, but I get:
```
[info] Linking (2383 ms)
[error] missing symbols:
[error] * M12java.net.URID5toURLL12java.net.URLEO
[error]   - from M34org.ekrich.config.impl.PlatformUriD5toURLL12java.net.URLEO
[error]   - from M33org.ekrich.config.impl.Parseable$D10relativeToL12java.net.URLL16java.lang.StringL12java.net.URLEO
[error]   - from M45org.ekrich.config.impl.Parseable$ParseableURLD10relativeToL16java.lang.StringL33org.ekrich.config.ConfigParseableEO
[error]   - from M33org.ekrich.config.impl.Parseable$D6newURLL12java.net.URLL36org.ekrich.config.ConfigParseOptionsL32org.ekrich.config.impl.ParseableEO
[error]   - from M32org.ekrich.config.ConfigFactory$D8parseURLL12java.net.URLL36org.ekrich.config.ConfigParseOptionsL24org.ekrich.config.ConfigEO
[error]   - from M38org.ekrich.config.impl.SimpleIncluder$D25includeURLWithoutFallbackL38org.ekrich.config.ConfigIncludeContextL12java.net.URLL30org.ekrich.config.ConfigObjectEO
```

and a bunch of other things :O

**Comment by tgodzik** [comment]
My current code is at https://github.com/scalameta/scalafmt/compare/master...tgodzik:add-scala-native?expand=1

There are still some other minor stuff to work out, but I am completely stumped by this one.

**Comment by ekrich** [comment]
I’ll have to look at that later today. I could have missed something there or something changed in Scala Native.

**Comment by ekrich** [comment]
I am not seeing anything obvious. Scala Native still has an empty `URL` class and the `sconfig` code seems ok. Do you mind trying `nativeLinkStubs := true` in the build. The `URI.toURL` method is a stub in Scala Native.
```Scala
  @scalanative.annotation.stub
  def toURL(): java.net.URL = ???
```

When working with Olafur previously, the code did not use the `ConfigFactory.parseFile` code path. Unfortunately, I was concentrating on having a smooth transition to Scala 3 and refactoring the tests for more code coverage was further down on the list for `sconfig`. The tested path was how the `scalafmt` code worked before for Scala Native and is shown here: https://github.com/ekrich/sconfig/blob/main/docs/SCALA-NATIVE.md We may be hitting an unknown code path or something else is going on. 

After looking further at this it looks like `PlatformUri` should not be shared with the JVM. We are just kicking the link stubs one layer deeper to more link stubs.

Wow, you had to bring the difflib across from `munit` too.

**Comment by tgodzik** [comment]
> Scala Native still has an empty `URL` class. Do you mind trying `nativeLinkStubs := true` in the build. The `URI.toURL` method is a stub in Scala Native.

That helped. I am down to two issuses:

```
[error] * M26java.text.SimpleDateFormatRL16java.lang.StringE
...
[error] * T26java.text.SimpleDateFormat
```

Is SimpleDateFormat implemented in sjavatime or do we need to add it there?

```
[error] * M38java.util.concurrent.ConcurrentHashMapRE
...
[error] * M40java.util.concurrent.LinkedBlockingDequeRE
...
```

The concurrent collections is something that I can probably replace.

```
[error] * M18scopt.OptionParserRL16java.lang.StringE
...
[error] * T15java.io.Console
```

No idea about console and scoopt yet.

> Wow, you had to bring the difflib across from `munit` too.

I might try to release it as a separate library.

**Comment by tgodzik** [comment]
Ok, got it all compiling at least for now, but getting an exception:
```
scala.NotImplementedError: an implementation is missing
	at java.lang.Throwable.fillInStackTrace(Unknown Source)
	at scala.Predef$.???(Unknown Source)
	at java.net.URI.toURL(Unknown Source)
	at org.ekrich.config.impl.PlatformUri.toURL(Unknown Source)
	at org.ekrich.config.impl.SimpleConfigOrigin$.newFile(Unknown Source)
	at org.ekrich.config.impl.Parseable$ParseableFile.createOrigin(Unknown Source)
	at org.ekrich.config.impl.Parseable.postConstruct(Unknown Source)
	at org.ekrich.config.impl.Parseable$.newFile(Unknown Source)
	at org.ekrich.config.ConfigFactory$.parseFile(Unknown Source)
	at org.ekrich.config.ConfigFactory$.parseFile(Unknown Source)
	at metaconfig.sconfig.SConfig2Class$.$anonfun$gimmeConfFromFile$1(Unknown Source)
	at metaconfig.sconfig.SConfig2Class$$$Lambda$2.apply(Unknown Source)
	at metaconfig.sconfig.SConfig2Class$.gimmeSafeConf(Unknown Source)
	at metaconfig.sconfig.SConfig2Class$.gimmeConfFromFile(Unknown Source)
	at metaconfig.sconfig.package$.metaconfig$sconfig$package$$$anonfun$sConfigMetaconfigParser$1(Unknown Source)
	at metaconfig.sconfig.package$$anonfun$1.fromInput(Unknown Source)
	at metaconfig.Input$InputImplicits.parse(Unknown Source)
	at metaconfig.Input$InputImplicits.$anonfun$parse$1(Unknown Source)
	at metaconfig.Input$InputImplicits$$Lambda$1.apply(Unknown Source)
	at scala.Option.fold(Unknown Source)
	at metaconfig.Input$InputImplicits.parse(Unknown Source)
	at org.scalafmt.config.Config$.$anonfun$hoconFileToConf$2(Unknown Source)
	at org.scalafmt.config.Config$$$Lambda$2.apply(Unknown Source)
	at metaconfig.Configured.andThen(Unknown Source)
	at org.scalafmt.config.Config$.hoconFileToConf(Unknown Source)
	at org.scalafmt.cli.CliOptions.$anonfun$hoconOpt$3(Unknown Source)
	at org.scalafmt.cli.CliOptions$$Lambda$8.apply(Unknown Source)
	at scala.Option.map(Unknown Source)
	at org.scalafmt.cli.CliOptions.hoconOpt$lzycompute(Unknown Source)
	at org.scalafmt.cli.CliOptions.getHoconValueOpt(Unknown Source)
	at org.scalafmt.cli.CliOptions.getVersionIfDifferent(Unknown Source)
	at org.scalafmt.cli.Cli.findRunner(Unknown Source)
	at org.scalafmt.cli.Cli.run(Unknown Source)
	at org.scalafmt.cli.Cli.mainWithOptions(Unknown Source)
	at org.scalafmt.cli.Cli$.main(Unknown Source)
	at <none>.main(Unknown Source)
	at <none>.__libc_start_main(Unknown Source)
	at <none>._start(Unknown Source)
```
@ekrich I think that's something we should fix in sconfig?

**Comment by ekrich** [comment]
@tgodzik if you could only call `ConfigFactory.parseString`, I think it would work. `parseFile` can read `URL`s from the file and `URL` is a mine field in Scala Native and Scala.js - no support basically. Please do something like this if possible.
```Scala
import java.nio.file.{Files, Paths}
val bytes = Files.readAllBytes(Paths.get("src/main/resources/myapp.conf"))
val configStr = new String(bytes)
val conf = ConfigFactory.parseString(configStr)
```
We don't have `java.text` support so maybe `java.util.Formatter`? 

I put together `sjavatime` as a stop gap until https://github.com/cquiroz/scala-java-time could be updated but nobody has managed to get this library updated yet.

Hope this helps.

**Comment by tgodzik** [comment]
I will do with `toString` for now, it's not that bad. Turns out though there is a bunch of other issues:
```
java.lang.UnsupportedOperationException
	at java.lang.Throwable.fillInStackTrace(Unknown Source)
	at java.nio.file.PathMatcherImpl$.apply(Unknown Source)
	at scala.scalanative.nio.fs.UnixFileSystem.getPathMatcher(Unknown Source)
	at org.scalafmt.config.ProjectFiles$FileMatcher$.$anonfun$nio$1(Unknown Source)
	at org.scalafmt.config.ProjectFiles$FileMatcher$$$Lambda$2.apply(Unknown Source)
	at scala.collection.immutable.List.map(Unknown Source)
	at scala.collection.immutable.List.map(Unknown Source)
	at org.scalafmt.config.ProjectFiles$FileMatcher$.create(Unknown Source)
	at org.scalafmt.config.ProjectFiles$FileMatcher$.nio(Unknown Source)
	at org.scalafmt.config.ProjectFiles$FileMatcher$.apply(Unknown Source)
	at org.scalafmt.cli.ScalafmtCoreRunner$.$anonfun$run$2(Unknown Source)
	at org.scalafmt.cli.ScalafmtCoreRunner$$$Lambda$2.apply(Unknown Source)
	at metaconfig.Configured$ConfiguredImplicit.fold(Unknown Source)
	at org.scalafmt.cli.ScalafmtCoreRunner$.run(Unknown Source)
	at org.scalafmt.cli.Cli.runWithRunner(Unknown Source)
	at org.scalafmt.cli.Cli.run(Unknown Source)
	at org.scalafmt.cli.Cli.mainWithOptions(Unknown Source)
	at org.scalafmt.cli.Cli$.main(Unknown Source)
	at <none>.main(Unknown Source)
	at <none>.__libc_start_main(Unknown Source)
	at <none>._start(Unknown Source)
```
Scalafmt uses `getPathMatcher` which seems to be not implemented and that is used in a bunch of places. I found some failing regexes, which I asked Wojciech to investigate.

I also found another issue with running git:
```
        sys.process
          .Process(cmd, workingDirectory.jfile)
          .!!(swallowStderr)
          .trim
```
Seems that it's not possible currently to run a process?

There is one last issue that deals with Threads, but I guess I can work around it for Scala Native.

**Comment by ekrich** [comment]
We do support `java.lang.Process` and `java.lang.ProcessBuilder`. I looked and it seems that `scala.sys.process.ProcessImpl` requires threads.

**Comment by tgodzik** [comment]
Ach ok, that fixed it. The remaining issues are:
- `scala.scalanative.nio.fs.UnixFileSystem.getPathMatcher` maybe there is a way to work around it
- failing regexes - I don't think there is a way around those currently
- non-thread implementation of `UpdateDisplayThread`

**Comment by ekrich** [comment]
What is the issue with `scala.scalanative.nio.fs.UnixFileSystem.getPathMatcher`?

cc/ Our regex expert @LeeTibbert

**Comment by tgodzik** [comment]
> What is the issue with `scala.scalanative.nio.fs.UnixFileSystem.getPathMatcher`?
> 
> cc/ Our regex expert @LeeTibbert

It's not currently implemented for Scala Native, we could try turning the glob patterns into regexes aka https://stackoverflow.com/questions/1247772/is-there-an-equivalent-of-java-util-regex-for-glob-type-patterns
but that is a bit hacky.

Anyway, I am trying out that solution (seems to do the trick for most cases), but another thing popped up after implementing that:
```
org.scalafmt.cli.FailedToFormat: /home/tgodzik/Documents/metals/tests/unit/src/test/scala/tests/worksheets/WorksheetNoDecorationsLspSuite.scala
        <no stack trace available>
Caused by: java.lang.NullPointerException
        at java.lang.Throwable.fillInStackTrace(Unknown Source)
        at scala.scalanative.runtime.package$.throwNullPointer(Unknown Source)
        at <none>.(Unknown Source)
        at org.scalafmt.internal.State$.getColumnsWithStripMargin(Unknown Source)
        at org.scalafmt.internal.State$.getColumns(Unknown Source)
        at org.scalafmt.internal.State.next(Unknown Source)
```

I pushed the current state to the branch if anyone wants to take a look.

**Comment by tgodzik** [comment]
Ok, we worked around the Regexes thanks to @sjrd - we might need to fix some issues with the workaround still - but everything finally compiles and runs!

I've put a draft PR here: https://github.com/tgodzik/scalafmt/pull/13/files

I still need to:
- Make TermDisplay work for Native
- Add tests and fix everything
- Tidy up

This might still be a bit of work. Next step would be the publishing story, I think we could try the same as in Bloop, publish artifact to github and link to them from coursier, which would allow us to install binaries using `cs`

**Comment by jchyb** [comment]
Some updates. I took over from @tgodzik to finish the PR and I mostly did, it's pretty much done (https://github.com/jchyb/scalafmt/pull/1). Testing on a my small unrelated repo yielded very promising results. However. I decided to do some real benchmarks on the scala-native repo itself, and... there are some problems. Some of the bigger files take extremely long to process (from about 1s sys time on JVM to 10s on Native). I have reasons to believe it's related to scala-native/scala-native#1713, but I need to investigate further. I don't think it makes much sense to merge before fixing the underlying issue.

**Comment by ekrich** [comment]
@jchyb I know @densh took a look at the performance issues recently and posted on the Scala Native discord.

I am wondering too if it might be a good idea to introduce changes made in the Scala Native PR branch here that can be merged. It seems reasonable to have the code support Scala Native even if the usability is sub par. I feel that runtime or other issues are somewhat tangential to this issue. It might make it easier for others to get involved as well.

Currently, 3 of the 5 check boxes are completed.

**Comment by jchyb** [comment]
@ekrich I have been meaning to get back to this. For now I would like to properly implement glob support on the SN side. Should not be too difficult and the regex wrapping system used before was not great and on further consideration I would not want to introduce too many SN-specific considerations for the scalafmt team. Additionally @kitbellew review (for which I am very grateful) revealed some issues with f.e. the lack of test cases there.

In terms of performance, I will also try testing scalafmt with the newest SN version and report my findings today in the issue. The recent changes to Strings seem substantial.

As for the check boxes they look slightly outdated, I was under the impression that sconfig was already used on the metaconfig side (but I don't remember exactly at this point)

**Comment by ekrich** [comment]
@jchyb Sounds like a plan - sconfig was merged for the Scala Native cross project as far as I know.

**Comment by ekrich** [comment]
@tgodzik So this can finally be closed? 🥰
cc @olafurpg

**Comment by LeeTibbert** [comment]
For the curious, what _is_ the state of scalafmt Scala Native support?  Is it still a goal?
Does it exist? 

As mentioned several times, Scala Native scalafmt, when a small group of us had it working,
was wicked fast formatting the entire SN GitHub repository tree.

I understand, all too well, that a project can not support all possibilities. Just trying to 
get back up to speed.

**Comment by kitbellew** [comment]
@LeeTibbert it should work although we haven't yet figured out all kinks with releasing.

also, i haven't replied to your question about regular expressions, my apologies. i will try to find some time and get my thoughts together.

**Comment by LeeTibbert** [comment]
kitbellew  Thank you for the update.

>  i will try to find some time and get my thoughts together.

When you figure out either or both parts, could you please teach me?
Of course, the "figuring out" your thoughts part is easier for me than
the reflective "figuring out" my own thoughts. ;-)

> it should work although we haven't yet figured out all kinks with releasing.

If, in the fullness of time, you need a field test guinea pig, please yell.
I use scalafmt a few hundred time a day.

**Comment by tgodzik** [comment]
The last release includes Scala Native binaries. I think we can finally close the issue!

**Comment by tgodzik** [comment]
The actual binaries are not connected to any distribution channels yet. They are available to download in the release. Scala Cli itself will use jvm for anything that doesn't have native available. In this case it's the 3.9.1 version which isn't yet beings used with the new artifacts.

So overall, the only way to test it currently is to download from the release page the relevant binary

**Comment by LeeTibbert** [comment]
@tgodzik 

Thank you for the timely & helpful response.  I was able to use your clues to get up and going.

I use Linux "file" command to show that I was indeed running a ELF binary.

Some rough data from running the equivalent of "scalafmt --check" on a scratch copy
of the Scala Native source tree.

Scala Native binary
```
$ time ./scalafmt-x86_64-pc-linux --check
Looking for unformatted files... [##########]
  out 100.0% 1876, fmt 100.0% 1876, in 100.0% 1876
All files are formatted with scalafmt :)

real    0m21.737s
user    1m1.900s
sys     0m1.137s
```
```
JVMtime env CHECK_MODIFIED_ONLY=true scripts/scalafmt

$ time env CHECK_MODIFIED_ONLY=true scripts/scalafmt
Reformatting... [##########]
  out 100.0% 1876, fmt 100.0% 1876, in 100.0% 1876

real    0m50.774s
user    3m5.157s
sys     0m1.705s
```

 I'll certainly take a reduction in real elapsed time to only 42% of original!

Good work by you and your crew.

I am in the middle of submitting a SN PR now. When I have a chance, I'll try
the a similar test on a 'dirty' directory where some files actually need formatting,
not just checking.  That should increase the I/O load and stress/exercise SN's
file & regex handling.  I fear what monsters are hiding in _that_ cupboard.

I'll also, when I slow down, file a scalafmt Issue about _something_ "-v" "--build-info" or 
other showing the "flavor" of the executing scalafmt.  Not always, but if one asks nicely.

After a 7 year journey, I am quite pleased how this is turning out. I hope you and
your crew are also.
